x-pack/plugins/security/server/session_management/session_management_service.ts

```diff
@@ -0,0 +1,160 @@
+ */
+
+import { Observable, Subscription } from 'rxjs';
+import {
+  HttpServiceSetup,
+  ILegacyClusterClient,
+  Logger,
+  SavedObjectsErrorHelpers,
+} from '../../../../../src/core/server';
+import { TaskManagerSetupContract, TaskManagerStartContract } from '../../../task_manager/server';
+import { ConfigType } from '../config';
+import { OnlineStatusRetryScheduler } from '../elasticsearch';
+import { SessionCookie } from './session_cookie';
+import { SessionIndex } from './session_index';
+import { Session } from './session';
+
+export interface SessionManagementServiceSetupParams {
+  readonly http: Pick<HttpServiceSetup, 'basePath' | 'createCookieSessionStorageFactory'>;
+  readonly config: ConfigType;
+  readonly clusterClient: ILegacyClusterClient;
+  readonly kibanaIndexName: string;
+  readonly taskManager: TaskManagerSetupContract;
+}
+
+export interface SessionManagementServiceStartParams {
+  readonly online$: Observable<OnlineStatusRetryScheduler>;
+  readonly taskManager: TaskManagerStartContract;
+}
+
+export interface SessionManagementServiceSetup {
+  readonly session: Session;
+}
+
+/**
+ * Name of the task that is periodically run and performs session index cleanup.
+ */
+export const SESSION_INDEX_CLEANUP_TASK_NAME = 'session_cleanup';
+
+/**
+ * Service responsible for the user session management.
+ */
+export class SessionManagementService {
+  private statusSubscription?: Subscription;
+  private sessionIndex!: SessionIndex;
+  private config!: ConfigType;
+
+  constructor(private readonly logger: Logger) {}
+
+  setup({
+    config,
+    clusterClient,
+    http,
+    kibanaIndexName,
+    taskManager,
+  }: SessionManagementServiceSetupParams): SessionManagementServiceSetup {
+    this.config = config;
+
+    const sessionCookie = new SessionCookie({
+      config,
+      createCookieSessionStorageFactory: http.createCookieSessionStorageFactory,
+      serverBasePath: http.basePath.serverBasePath || '/',
+      logger: this.logger.get('cookie'),
+    });
+
+    this.sessionIndex = new SessionIndex({
+      config,
+      clusterClient,
+      kibanaIndexName,
+      logger: this.logger.get('index'),
+    });
+
+    // Register task that will perform periodic session index cleanup.
+    taskManager.registerTaskDefinitions({
+      [SESSION_INDEX_CLEANUP_TASK_NAME]: {
+        title: 'Cleanup expired or invalid user sessions',
+        type: SESSION_INDEX_CLEANUP_TASK_NAME,
+        createTaskRunner: () => ({ run: () => this.sessionIndex.cleanUp() }),
+      },
+    });
+
+    return {
+      session: new Session({
+        logger: this.logger,
+        sessionCookie,
+        sessionIndex: this.sessionIndex,
+        config,
+      }),
+    };
+  }
+
+  start({ online$, taskManager }: SessionManagementServiceStartParams) {
+    this.statusSubscription = online$.subscribe(async ({ scheduleRetry }) => {
+      try {
+        await Promise.all([this.sessionIndex.initialize(), this.scheduleCleanupTask(taskManager)]);
```

Comment:
__note:__ Chatted with the Alerting team that owns Task Manager and the common understanding is that removing task manager index or its content without Kibana restart is considered as a "critical misuse" and we shouldn't bother supporting it. It's not expensive for us to be more resilient (that what we discussed and I did in the last commit), but Kibana may still not be usable as other Task Manager consumers aren't handing this.

While dealing with this I realized that `ensureScheduled` just checks that task with such ID exists and that's it. It doesn't detect that parameters of the task may have changed. This may be a problem for us since interval is configurable (that's why if you set interval to `1m` and even restart Kibana, interval will still be default `1h` until you clear task manager index). I fixed that as well.